go/kbfs/simplefs/uploads.go

```diff
@@ -0,0 +1,250 @@
+	u, ok := m.uploads[uploadID]
+	return u, ok
+}
+
+func (m *uploadManager) waitForDownload(uploadID string) {
+	upload, ok := m.getUpload(uploadID)
+	if !ok {
+		return
+	}
+	defer m.publisher.PublishChange(keybase1.SubscriptionTopic_UPLOAD_STATUS)
+
+	err := m.k.SimpleFSWait(m.makeContext(), upload.opid)
+	if err != nil {
+		m.doWhileLocked(func() {
+			upload, ok := m.uploads[uploadID]
+			if !ok {
+				return
+			}
+			if errors.Cause(err) == context.Canceled {
+				upload.state.Canceled = true
+			} else {
+				errStr := err.Error()
+				upload.state.Error = &errStr
+			}
+			m.uploads[uploadID] = upload
+		})
+		return
+	}
+	if upload.dirToDelete != nil {
+		os.RemoveAll(*upload.dirToDelete)
+	}
+	m.doWhileLocked(func() {
+		delete(m.uploads, uploadID)
+	})
+}
+
+const uploadSuffixMax = 1024
+
+func (m *uploadManager) _start(ctx context.Context,
+	sourceLocalPath string, parentPath string) (opid keybase1.OpID, dstPath keybase1.KBFSPath, err error) {
+	opid, err = m.k.SimpleFSMakeOpid(ctx)
+	if err != nil {
+		return keybase1.OpID{}, keybase1.KBFSPath{}, err
+	}
+	basename := filepath.Base(sourceLocalPath)
+
+renameLoop:
+	for i := 0; i < uploadSuffixMax; i++ {
+		name := basename
+		if i > 0 {
+			name = fmt.Sprintf("%s (%d)", basename, i)
+		}
+		dstPath = keybase1.NewPathWithKbfsPath(
+			path.Join(parentPath, name)).Kbfs()
+
+		// First check with stat. This should cover most cases, and is
+		// the last resort for avoiding merging directories where we
+		// don't get something like O_CREATE for free.
+		if _, err = m.k.SimpleFSStat(ctx, keybase1.SimpleFSStatArg{
+			Path: keybase1.NewPathWithKbfs(dstPath),
+		}); err == nil {
+			continue renameLoop
+		}
+
+		err = m.k.SimpleFSCopyRecursive(ctx, keybase1.SimpleFSCopyRecursiveArg{
+			OpID:                   opid,
+			Src:                    keybase1.NewPathWithLocal(sourceLocalPath),
+			Dest:                   keybase1.NewPathWithKbfs(dstPath),
+			OverwriteExistingFiles: false,
+		})
+		switch errors.Cause(err) {
+		case os.ErrExist:
+			continue renameLoop
+		case nil:
+			return opid, dstPath, nil
+		default:
+			return keybase1.OpID{}, keybase1.KBFSPath{}, err
+		}
+	}
+
+	return keybase1.OpID{}, keybase1.KBFSPath{},
+		errors.New("too many rename attempts")
+}
+
+func (m *uploadManager) getDirToDeleteRLocked(sourceLocalPath string) *string {
+	sourceParent := filepath.Dir(sourceLocalPath)
+	if m.tempPaths[sourceParent] {
+		return &sourceParent
+	}
+	return nil
+}
+
+func (m *uploadManager) start(ctx context.Context, sourceLocalPath string,
+	targetParentPath keybase1.KBFSPath) (uploadID string, err error) {
+	uploadID = strconv.FormatInt(time.Now().UnixNano(), 16)
+	opid, dstPath, err := m._start(ctx, sourceLocalPath, targetParentPath.Path)
+	if err != nil {
+		return "", err
+	}
+
+	defer m.publisher.PublishChange(keybase1.SubscriptionTopic_UPLOAD_STATUS)
+
+	m.doWhileLocked(func() {
+		m.uploads[uploadID] = upload{
+			dirToDelete: m.getDirToDeleteRLocked(sourceLocalPath),
```

Comment:
Nit: this is the only place this function is called, and it's called while write-locked, not read-locked, which is a little confusing.